### ISSUE TITLE
Apache/PHP virtual memory leak with 0.26.0

### ISSUE THREAD

**Issue by ctrlrsf** [DataDog/dd-trace-php#465]
Hi,

We started seeing an apache/PHP virtual memory leak after upgrading to 0.26.0. With 0.25.0 apache2's root process **virtual** memory usage remains constant for hours/days. With 0.26.0, you can immediately see it steadily climb until it runs out of VM (as high as 25 or 40+GB, depending on server) and it can no longer fork a new process.

In screenshot below from Datadog Process Monitor, bottom steady red lines are several instances running 0.25.0. Top red diagonal line is a server running 0.26.0. Yellow line is a server where datadog-php-tracer was upgraded from 0.25.0 to 0.26.0. 

![image](https://user-images.githubusercontent.com/6004738/58892093-a19df200-86bb-11e9-9fd4-16fe8330d73e.png)

After several hours we'll start to see these errors and Apache/PHP will hang:
```
Jun  4 12:34:01 xyz apache-error: [Tue Jun 04 12:34:01.029948 2019] [mpm_prefork:error] [pid 19722] (12)Cannot allocate memory: AH00159: fork: Unable to fork new process
Jun  4 12:34:11 xyz apache-error: [Tue Jun 04 12:34:11.030127 2019] [mpm_prefork:error] [pid 19722] (12)Cannot allocate memory: AH00159: fork: Unable to fork new process
```

VSZ will continue to grow every few minutes. RSS remains about the same.
```
sdf@redacted:~$ ps aux | grep root.*apache2
USER       PID %CPU %MEM    VSZ   RSS TTY      STAT START   TIME COMMAND
root      3796  0.0  0.3 1593140 49980 ?       Ssl  15:20   0:00 /usr/sbin/apache2 -k start
```
Versions:
apache2: 2.4.18-2ubuntu3.10
php7.2: 7.2.19

Seems related to PR https://github.com/DataDog/dd-trace-php/pull/450 that's in the 0.26.0 release.

Update: confirmed downgrading to 0.25.0 keeps virtual memory size stable:
![image](https://user-images.githubusercontent.com/6004738/58894059-4a9a1c00-86bf-11e9-8804-56e380c1b650.png)

vs 0.26.0:

![image](https://user-images.githubusercontent.com/6004738/58894006-33f3c500-86bf-11e9-974d-a51a30dc9838.png)

**Comment by pawelchcki** [comment]
Hey @ctrlrsf
Thanks for reporting it!
It seems our new background thread for sending traces has some major leak.

Its double troubling since we've so far were relying on automated valgrind tests to catch these kind of problems in development, but for some reason they haven't tripped on this very obvious (in retrospect) case!

The fix for the leak will be upcoming shorty, and we're going to double check our memory leak detection and its results to make sure this doesn't happen anymore in the future  and that we don't have any other memory leaks).

**Comment by pawelchcki** [comment]
We've just released version `0.27.1` which contains fix for this leak.

**Comment by ctrlrsf** [comment]
Thanks for the quick fix @pawelchcki! We'll give 0.27.1 a shot and report back any problems.

**Comment by ctrlrsf** [comment]
Been running 0.27.1 overnight on canary and apache VM usage is stable. Thanks again for the quick resolution.

![image](https://user-images.githubusercontent.com/6004738/59040795-6fb89700-8845-11e9-95b4-b6d54ff559ec.png)

**Comment by kjdev** [comment]
![dd](https://user-images.githubusercontent.com/465132/59891898-fd81ae00-9411-11e9-9d45-60ac81524a52.png)

``` sh
$ docker build --build-arg DD_APM_VERSION=0.28.0 -t dd:0.28.0 .
$ docker build --build-arg DD_APM_VERSION=0.26.0 -t dd:0.26.0 .
$ docker build --build-arg DD_APM_VERSION=0.25.0 -t dd:0.25.0 .
$ docker-compose up
$ echo -e "http://localhost:8888/\nhttp://localhost:8886/\nhttp://localhost:8885/"> urls.txt
$ for i in {1..10} ; do siege -c 30 -r 10000 -b -f urls.txt; sleep 30; done
```

Memory Usage: `docker stats --no-stream --format "{{.Container}} {{.MemUsage}}"`

> Dockerfile
> ``` Dockerfile
> FROM alpine:3.9
> RUN apk --no-cache upgrade && apk --no-cache add curl php7-opcache php7-apache2
> COPY app.conf /etc/apache2/conf.d/app.conf
> COPY index.php /app/index.php
> ARG DD_APM_VERSION=0.28.0
> RUN ( curl -LO https://github.com/DataDog/dd-trace-php/releases/download/${DD_APM_VERSION}/datadog-php-tracer_${DD_APM_VERSION}_noarch.apk && apk add datadog-php-tracer_${DD_APM_VERSION}_noarch.apk --allow-untrusted ) && rm -f datadog-php-tracer_${DD_APM_VERSION}_noarch.apk
> EXPOSE 80
> ENTRYPOINT ["/usr/sbin/httpd"]
> CMD ["-D", "FOREGROUND"]
> ```
> app.conf
> ``` app.conf
> <VirtualHost *:80>
>   DocumentRoot /app
>   DirectoryIndex index.php
>   <Directory /app>
>     Options FollowSymLinks
>     AllowOverride All
>     Require all granted
>   </Directory>
> </VirtualHost>
> ```
> index.php:
> ``` index.php
> <?php
> echo 'OK';
> ```
> docker-compose.yml:
> ``` docker-compose.yml
> version: '3.7'
> services:
>   dd-28:
>     image: dd:0.28.0
>     ports:
>       - 8888:80
>   dd-26:
>     image: dd:0.26.0
>     ports:
>       - 8886:80
>   dd-25:
>     image: dd:0.25.0
>     ports:
>       - 8885:80
> ```

**Comment by labbati** [comment]
Thanks @kjdev for reporting that, let us work on this. We will make sure to improve our CI pipeline to make sure that, if confirmed, such increases in memory consumption will not pass CI.

**Comment by kjdev** [comment]
![dd](https://user-images.githubusercontent.com/465132/62742094-39cba500-ba78-11e9-93bd-9da4c1ab11cf.png)

It seems to have been resolved in 0.30.0.

**Comment by labbati** [comment]
Hi @kjdev, thanks for confirming that 0.30.0 fixes what you observed.
